src/openforms/formio/datastructures.py

```diff
@@ -84,3 +86,38 @@ def is_visible_in_frontend(self, key: str, values: DataMapping) -> bool:
             component = cast(Component, glom(self.configuration, path))
             nodes.append(component)
         return all(is_visible_in_frontend(node, values) for node in nodes)
+
+
+class FormioData(UserDict):
+    """
+    Handle formio (submission) data transparently.
+
+    Form.io supports component keys in the format 'topLevel.nested' which get converted
+    to deep-setting of object properties (using ``lodash.set`` internally). This
+    datastructure mimicks that interface in Python so we can more naturally perform
+    operations like:
+
+    .. code-block:: python
+
+        data = FormioData()
+        for component in iter_components(...):
+            data[component["key"]] = ...
+
+    without having to worry about potential deep assignments or leak implementation
+    details (such as using ``glom`` for this).
+    """
+
+    data: dict[str, JSONValue]
+
+    def __getitem__(self, key: Hashable):
+        return cast(JSONValue, glom(self.data, key))
```

Comment:
I *know* we don't like the casts, but the premise really is json-serializable data when working with formio submission data. Though the datatypes thing might complicate this a bit more?